max_gap_fix

daniel-caichac-DHI opened the pull request:
Hi @jsmariegaard / @ecomodeller , I need help :(
I am trying to implement a `max_gap` feature when interpolating `Model dt` to `Measurement dt` to avoid unrealistic interpolations as the following
![image](https://user-images.githubusercontent.com/97288080/205649417-e81d2536-74bf-4f92-ba70-03aad302d3d1.png)
![image](https://user-images.githubusercontent.com/97288080/205649440-07338e43-03d9-4723-b45d-be75d2212cf9.png)

It works when I hard-code a max dt with a given value (eg, if my dt is 10min and I want a max interpolation of gap of 1 hour, then when I set `limit=6` in here it works ok)
https://github.com/DHI/fmskill/blob/74f15e9e5b0ad1116f4a842297e0010c635bbe7d/fmskill/comparison.py#L1173-L1177
but then I tried to implement this `max_gap` as a user input and failed miserably. I got lost with all the child classes and inheritances.

Could you please help me getting this one working? I know it is not far away from being there.

Hendrik1987 commented:
#147

ecomodeller commented:
I have changed the code to put the `max_gap` parameter on the `Comparers`. The user will supply this parameter in the call to `.extract(max_gap=...)`

Please go ahead and include a test file where using this parameter makes sense.

daniel-caichac-DHI commented:
@ecomodeller I added a new line in the case `max_gap < measurement dt`  (in which case a warning is prompted and no model interpolation is performed), and I also added a connection test with a new `test_file.dfsu` with gaps and and a new test that connects it with and without a specified `max_gap`
I think we are good to merge now.

Hendrik1987 commented:
@ecomodeller and @daniel-caichac-DHI , I have an example and test on this PR #147. Should I bring it along to here? It will still fail, as it uses `.compare()`, which does not (yet) pass `max_gap` to `.extract()`.
Do we need to add this or are we happy as it is?

Hendrik1987 commented:
@daniel-caichac-DHI : Are you sure your `.dfsu` file has a gap? I have been there before and then learned `.dfsu` files cannot be non-equidistant. I.e., the gap scenario is limited to `.dfs0`, `df`, `.grib` and `.nc` model results.

daniel-caichac-DHI commented:
@Hendrik1987  I tested with the dfsu with nans for the same reasons you mentioned, but I also tested (maybe I should add another file to the test files? it just seemed redundant) a dfs0 with gaps (non-eq dfs0) and also worked. 
Maybe I will do it now to clear doubts

Hendrik1987 commented:
@daniel-caichac-DHI : give me a moment, just bringing the other tests along...

jsmariegaard commented:
I still think it should be defined on the ModelResult though :-) 

    mr = fmskill.ModelResult(fn, max_gap=1800)

daniel-caichac-DHI commented:
I am updating with a non-eq test that it does not fail

Hendrik1987 commented:
Added test for a `df` model result and `.compare()`. I agree with @jsmariegaard 's argument that `max_gap` is a model property, however, `.compare()` does not require to define a `ModelResult`...

daniel-caichac-DHI commented:
with the new test the current implementations fails though , as it is right now implemented only in `extract` not in `compare`

daniel-caichac-DHI commented:
@jsmariegaard  I moved the `max_gap` from the `extraction` to `model result`  and I tested it and it works eg with

`mr=ModelResult(r'file.dfs0')`
`mr=ModelResult(r'file.dfs0',max_gap=3600)`

it also works with pandas dataframe as model results and I think xarray. The problem I have is that moving this property to `model result` now makes me fail all of the multi model compare tests (so I cannot even do `cc.skill()` if I have more than 1 model in my connector).
I honestly do not know how to fix this,  I kind of need help with this.

daniel-caichac-DHI commented:
@jsmariegaard  
I changed the `max_gap_tests`  because they were made by Hendrik before we built the `max_gap` as a property of a `ModelResult`

Now the tests should be succesful

jsmariegaard commented:
I agree that we are close to having something which works as intended. But it required a lot of changes all over the code and that "smells" to me. I am not convinced that this is the right way to deal with this problem (requested feature). I know it was my own idea to add "max_gap" as an argument to the ModelResult, but I am no longer sure this is the right way. Also, it makes me worried that _interp_df() has grown from a few lines to almost a whole page. I think the max_gap feature is too small to allow for such a substantial increase in complexity.

daniel-caichac-DHI commented:
We can always go back to like 7 commits ago where the `max_gap` was in the connector, and we could just rename it to `max_model_interp_span` or something like that.
The code was much cleaner then
https://github.com/DHI/fmskill/pull/148/commits/95cd56bda00a2550eab32b2e42f544e56957ec50

Hendrik1987 commented on fmskill/comparison.py:
```diff
@@ -1149,22 +1150,65 @@ def __copy__(self):
+            vals, counts = np.unique(ix, return_counts=True)
+            mode_value_ix = np.argwhere(counts == np.max(counts))
+            dt_meas = vals[mode_value_ix][0][0] / np.timedelta64(1, "s")
+            limit = int(max_gap / dt_meas)
```
Is this pandas solution actually robust enough? A fix limit? `dt_meas` will often be non-equidistant, right?

Hendrik1987 commented on tests/test_trackcompare.py:
```diff
@@ -100,6 +100,23 @@ def test_extract_no_spatial_overlap_dfs0(modelresult, observation_df):
     assert len(cc.all_df) == 0
 
 
+# def test_extract_from_model_result_with_nan(sw_Hm0_df, Hm0_EPL):
```
Why in `test_trackcompare`?

jsmariegaard replied:
That was probably a mistake - my primary concern was to move it out of ModelResult :-)

Hendrik1987 commented:
@daniel-caichac-DHI and @jsmariegaard , I did not actively contribute here, so to not mess with your code, I added a temporary [test notebook](https://github.com/DHI/fmskill/blob/10e3222de32c4e280a9f6f7654af2ac2f22fec27/notebooks/temp_test_df_interpolation.ipynb) with some thoughts. In short:
1. I think using `pd.interpolate(..., limit)` might not be robust enough (_v2_ in notebook)
2. I think better would be to identify valid observation time steps to which the models can be interpolated. Our internal tools have an elegant solution (@patrickgrode ) (_v3_ in notebook, work in progress).
3. Dropping observation data points where not all models are available is a seperate / last step (_misc_ in the nootebook)

jsmariegaard commented:
Thanks for the notebook, @Hendrik1987, would be great to reformulate it as a number of unit tests for the interp_df method 👍

Hendrik1987 commented:
Sure @jsmariegaard , where should the tests be placed?
And what are your thoughts on `limit` (v2) vs. `valid_query_times` (v3). If you approve I can work on the latter, too.

Hendrik1987 commented on fmskill/connection.py:
```diff
@@ -295,7 +296,7 @@ def _parse_observation(self, obs) -> PointObservation:
             raise ValueError(f"Unknown observation type {type(obs)}")
 
-    def extract(self) -> PointComparer:
+    def extract(self, max_model_gap: float=None) -> PointComparer:
```
@daniel-caichac-DHI : `.extract()` should take `max_model_gap`, no?

daniel-caichac-DHI replied:
I don't know, I had a script and tested with this commit and `max_gap=3600`  in `extract()` and it works,
https://github.com/DHI/fmskill/pull/148/commits/95cd56bda00a2550eab32b2e42f544e56957ec50
then I tested with the most recent version and `max_model_gap=3600`  in `extract()` and it does not connect anything :(

jsmariegaard replied:
Could you maybe add a test that illustrates the problem? expected vs actual...

daniel-caichac-DHI replied:
* When I don't use max gap I get this, 278 points
![image](https://user-images.githubusercontent.com/97288080/210729117-4065d6bb-1ab4-4b18-b31f-05caf9363c24.png)

* When use the old commit (https://github.com/DHI/fmskill/commit/95cd56bda00a2550eab32b2e42f544e56957ec50 ) I get this, 28 points
![image](https://user-images.githubusercontent.com/97288080/210729276-6f4d0bd2-56f8-4c58-9336-fa7b4941c10d.png)

* When I use the lastest commit of this branch , I get an error
![image](https://user-images.githubusercontent.com/97288080/210729426-be09b8e1-ac24-4ed2-a3f4-865628030f1d.png)

jsmariegaard replied:
It seems that the Connector misses this argument - only the specialized connectors, like PointConnector and TrackConnector has it.

daniel-caichac-DHI replied:
I just added a failing test where  `connector.extract(max_ga_mogel_gap=3600)` does not work.

Hendrik1987 commented:
@jsmariegaard : You asked me to reformulate notebook as couple of unit tests. I am generating test data within the notebook with a random component. Is it OK to save those as a few extra .dfs0 as additional test data? Maybe it is getting a bit messy with the test data, but I don't think we have an equivalent yet.

jsmariegaard commented:
> @jsmariegaard : You asked me to reformulate notebook as couple of unit tests. I am generating test data within the notebook with a random component. Is it OK to save those as a few extra .dfs0 as additional test data? Maybe it is getting a bit messy with the test data, but I don't think we have an equivalent yet.

Please do add more test data preferably with descriptive names 👍

Hendrik1987 commented:
> Please do add more test data preferably with descriptive names 👍

Cool, will do! You mean something like `model_final_new_v2_latest.dfs0`, right? ;-)

ecomodeller commented on tests/test_connector.py:
```diff
@@ -66,7 +66,7 @@ def o2():
     fn = "tests/testdata/SW/eur_Hm0.dfs0"
     obs = mikeio.read(fn, items=0).to_dataframe().rename(columns=dict(Hm0="obs")) + 1
-    rnd = np.random.random(len(obs.index))
+    rnd = np.random.random(len(obs.index)) - 1
```
How about setting the random seed?

Hendrik1987 commented:
Thanks @jsmariegaard for picking this up and sorry for not contributing more actively here!

jsmariegaard commented:
> Thanks @jsmariegaard for picking this up and sorry for not contributing more actively here!

No worries, @Hendrik1987, I really appreciated the work you did on the notebook! It was a great help finishing this PR.

jsmariegaard reviewed (approved):
Reviewed together with @ecomodeller IRL
